src/js/common/schemaform/FileField.jsx

```diff
@@ -0,0 +1,161 @@
+import PropTypes from 'prop-types';
+import React from 'react';
+import _ from 'lodash/fp';
+// import Scroll from 'react-scroll';
+import { displayFileSize } from '../../common/utils/helpers';
+import environment from '../../common/helpers/environment';
+
+export default class FileField extends React.Component {
+  onAddFile = (e) => {
+    const files = this.props.formData || [];
+    this.props.onChange(files.concat({
+      uploading: true
+    }));
+    this.uploadFile(e.target.files[0])
+      .then(fileInfo => {
+        const uploadingFiles = this.props.formData;
+        this.props.onChange(_.set(uploadingFiles.length - 1, {
+          errorMessage: null,
+          uploading: false,
+          fileName: fileInfo.fileName,
+          maxSize: fileInfo.maxSize,
+          confirmationNumber: fileInfo.confirmationNumber
+        }, uploadingFiles));
+      })
+      .catch(message => {
+        const uploadingFiles = this.props.formData || [];
+        this.props.onBlur(this.props.idSchema.$id);
+        this.props.onChange(_.set(uploadingFiles.length - 1, {
+          uploading: false,
+          errorMessage: message
+        }, uploadingFiles));
+      });
+  }
+
+  uploadFile = (file) => {
+    const uiSchema = this.props.uiSchema;
+
+    if (file.size > uiSchema['ui:options'].maxSize) {
+      return Promise.reject('File is too large to be uploaded');
+    }
+
+    const payload = new FormData();
+    payload.append('upload', file);
+
+    return fetch(`${environment.API_URL}${uiSchema['ui:options'].endpoint}`, {
+      method: 'POST',
+      headers: {
+        'X-Key-Inflection': 'camel'
+      },
+      body: payload
+    }).then(resp => {
+      if (resp.ok) {
+        return resp.json();
+      }
+
+      return Promise.reject(resp.statusText);
+    }, error => {
+      return Promise.reject(error.message);
+    });
+  }
+
+  removeFile = (index) => {
+    const newFileList = this.props.formData.filter((file, idx) => index !== idx);
```

Comment:
Not that it matters much given the tiny numbers we'll be dealing with here, but I was curious, so I did a jsperf and found this:
![image](https://user-images.githubusercontent.com/12970166/26907322-bb070494-4ba7-11e7-9518-d1e985ce55e9.png)
So for what it's worth, if we're dealing with large arrays and we need to return a clone of the array with only one index removed, slice and splice is the fastest. Good to know! (Eventually....maybe...?)